File: parallel.py
import sys
from math import atan2, sqrt
from time import time

import numexpr as ne
import numpy as np

N = 100_000_000

np.random.seed()

X = np.random.randn(N).tolist()
Y = np.random.randn(N).tolist()

working_set_GB = 3 * N * 8 / 2 ** 30

print(f'Python version: {sys.version}')

t0 = time()
Z0 = [sqrt(x * x + y * y) for x, y in zip(X, Y)]
t1 = time()
gbs = working_set_GB / (t1 - t0)
print(f'Time for an algebraic expression:     {t1 - t0:.3f} s / {gbs:.3f} GB/s')

t0 = time()
Z1 = [atan2(y, x) for x, y in zip(X, Y)]
t1 = time()
gbs = working_set_GB / (t1 - t0)
print(f'Time for a transcendental expression: {t1 - t0:.3f} s / {gbs:.3f} GB/s')

X = np.array(X)
Y = np.array(Y)
Z0 = np.array(Z0)
Z1 = np.array(Z1)

print(f'NumPy version: {np.__version__}')

t0 = time()
Z = np.sqrt(X * X + Y * Y)
t1 = time()
gbs = working_set_GB / (t1 - t0)
print(f'Time for an algebraic expression:     {t1 - t0:.3f} s / {gbs:.3f} GB/s')
np.testing.assert_equal(Z, Z0)

t0 = time()
Z = np.arctan2(Y, X)
t1 = time()
gbs = working_set_GB / (t1 - t0)
print(f'Time for a transcendental expression: {t1 - t0:.3f} s / {gbs:.3f} GB/s')
np.testing.assert_equal(Z, Z1)

print(f'Numexpr version: {ne.__version__}')

t0 = time()
ne.evaluate('sqrt(X * X + Y * Y)', out=Z)
t1 = time()
gbs = working_set_GB / (t1 - t0)
print(f'Time for an algebraic expression:     {t1 - t0:.3f} s / {gbs:.3f} GB/s')
np.testing.assert_equal(Z, Z0)

t0 = time()
ne.evaluate('arctan2(Y, X)', out=Z)
t1 = time()
gbs = working_set_GB / (t1 - t0)
print(f'Time for a transcendental expression: {t1 - t0:.3f} s / {gbs:.3f} GB/s')
np.testing.assert_equal(Z, Z1)

File: parallel.cpp
#include <cassert>
#include <chrono>
#include <execution>
#include <iostream>
#include <random>
#include <vector>

using namespace std;

int main() {
    constexpr size_t N{100'000'000};
    constexpr double working_set_GB{3.0 * N * sizeof(double) / exp2(30)};

    random_device rd;
    mt19937 gen{rd()};
    normal_distribution dis;

    vector<double> X(N), Y(N), Z(N), Z0(N), Z1(N);

    for (auto &&x:X) x = dis(gen);
    for (auto &&y:Y) y = dis(gen);

    cout << fixed;
    cout.precision(3);

    cout << "for loop version: \n";

    {
        auto t0 = chrono::high_resolution_clock::now();
        for (size_t i = 0; i < N; ++i) {
            Z0[i] = sqrt(X[i] * X[i] + Y[i] * Y[i]);
        }
        auto t1 = chrono::high_resolution_clock::now();
        chrono::duration<double> s = t1 - t0;
        auto gbs = working_set_GB / s.count();
        cout << "Time for an algebraic expression:     "
             << s.count() << " s / " << gbs << " GB/s\n";
    }

    {
        auto t0 = chrono::high_resolution_clock::now();
        for (size_t i = 0; i < N; ++i) {
            Z1[i] = atan2(Y[i], X[i]);
        }
        auto t1 = chrono::high_resolution_clock::now();
        chrono::duration<double> s = t1 - t0;
        auto gbs = working_set_GB / s.count();
        cout << "Time for a transcendental expression: "
             << s.count() << " s / " << gbs << " GB/s\n";
    }

    cout << "STL algorithms version: \n";

    {
        auto t0 = chrono::high_resolution_clock::now();
        transform(begin(X), end(X), begin(Y), begin(Z),
                  [](auto x, auto y) { return sqrt(x * x + y * y); });
        auto t1 = chrono::high_resolution_clock::now();
        chrono::duration<double> s = t1 - t0;
        auto gbs = working_set_GB / s.count();
        cout << "Time for an algebraic expression:     "
             << s.count() << " s / " << gbs << " GB/s\n";
        assert(Z == Z0);
    }

    {
        auto t0 = chrono::high_resolution_clock::now();
        transform(begin(X), end(X), begin(Y), begin(Z),
                  [](auto x, auto y) { return atan2(y, x); });
        auto t1 = chrono::high_resolution_clock::now();
        chrono::duration<double> s = t1 - t0;
        auto gbs = working_set_GB / s.count();
        cout << "Time for a transcendental expression: "
             << s.count() << " s / " << gbs << " GB/s\n";
        assert(Z == Z1);
    }

    cout << "Parallel STL algorithms version:\n";

    {
        auto t0 = chrono::high_resolution_clock::now();
        transform(execution::par, begin(X), end(X), begin(Y), begin(Z),
                  [](auto x, auto y) { return sqrt(x * x + y * y); });
        auto t1 = chrono::high_resolution_clock::now();
        chrono::duration<double> s = t1 - t0;
        auto gbs = working_set_GB / s.count();
        cout << "Time for an algebraic expression:     "
             << s.count() << " s / " << gbs << " GB/s\n";
        assert(Z == Z0);
    }

    {
        auto t0 = chrono::high_resolution_clock::now();
        transform(execution::par, begin(X), end(X), begin(Y), begin(Z),
                  [](auto x, auto y) { return atan2(y, x); });
        auto t1 = chrono::high_resolution_clock::now();
        chrono::duration<double> s = t1 - t0;
        auto gbs = working_set_GB / s.count();
        cout << "Time for a transcendental expression: "
             << s.count() << " s / " << gbs << " GB/s\n";
        assert(Z == Z1);
    }
}
